Lets deprecate

ScottSoren opened the pull request:
I expect this is my final PR on the way to ixdat v0.2.0 release. 

I'm requesting eyes from

-  @KennethNielsen for the use of deprecation decorators and changelog workflow, and 
-  @AnnaWiniwarter to see if the API changes are adequately described for use in e.g. EC-MS calibration, since I'm claiming it resolves the Issues related to that (including #20, where what had last time given an error will now give a `DeprecationWarning`).

The PR does the following:

- Introduces a changelog system. The instructions are described in docs/developing and at the top of NEXT_CHANGES.rst
- Describes the changes to the API going from ixdat v0.1.x to ixdat v0.2.0 in NEXT_CHANGES.rst. I will move this, after your review, to CHANGES.rst for release of v0.2.0. 
- Implements deprecation (introduced in #43). These are all set to give a warning in ixdat v0.2.x and raise an error in ixdat v0.3. The decorators are extensive so that people can for the most part keep running their scripts built in ixdat v0.1.x with . I think I've mentioned all the deprecations in NEXT_CHANGES.rst
- Fixes a bug by which it was sometimes impossible to look up a cache'd `series` in a measurement using its `series.name` as the key. 

In addition to the code here, I've updated all the other repositories that use ixdat so that they are fully ready for ixdat v0.2.0 release. In each of the following, I've made an **ixdat_v0p1** branch and merged the **ixdat_v0p2** branch into **main**. The scripts on the **ixdat_v0p1** branches are able to run with both ixdat v0.1.9 and v0.2.0 (code in this PR), with a lot of deprecation errors issued in the later. The README on the **main** branches describe the requirement of ixdat v0.2+ and have links to the **ixdat_v0p1**. This applies for:

- The tutorials repository: https://github.com/ixdat/tutorials
- The EC-MS article repository at https://github.com/ScottSoren/Huang2021
- The EC-MS article repository at https://github.com/ScottSoren/pyCOox_public

Looking forward to your feedback and to saying goodbye to ixdat v 0.1.x ! :)

KennethNielsen commented:
Looks like a lot of great work and important work. I will review shortly.

The tests fails due to a odd dependency problem I encountered during #46. The development dependencies list `setuputils`, but that package is a Python 2 only dependency, which doesn't make any sense for ixdat. My best guess is that at the `requirements-dev.txt` was created in retrospect in an environment that was already setup and working and therefore this just registered as a weird PyPI error when upgrading. In any case, I changed to dependency to `setuputils3` (which is the Python 3 versin of the former), which is my best guess at what it was supposed to be. So if you merge #46 and merge upstream/main into this, it should pass the tests (except for any style issues possibly introduced in this PR).

ScottSoren commented:
After merging in `main` and making a few tiny fixes, it passes the github CI tools :)

AnnaWiniwarter commented on NEXT_CHANGES.rst:
```diff
@@ -0,0 +1,102 @@
+
+- Renamed keyward in ``MSMeasurement.plot_flux()``:
+
+  - ``remove_backround`` (replaces ``removebackground``
```
it should be ``remove_background``, right? Is the typo also in the code?

KennethNielsen commented on src/ixdat/measurements.py:
```diff
@@ -439,6 +440,26 @@ def c_ids(self):
+    @deprecate(
+        "0.1",
+        "At present, ixdat measurements have a `calibration_list` but no compound "
+        "`calibration`, and the property just returns the first from the list.",
```
maybe ".. and **this** property"

KennethNielsen commented on src/ixdat/measurements.py:
```diff
@@ -439,6 +440,26 @@ def c_ids(self):
         self._calibration_list = [calibration] + self._calibration_list
 
+    @property
+    @deprecate(
```
Alright so, I was sort of expecting a problem around writing the keyword arguments out in full or not when I wrote it. I think, that since the is breaching multiple lines any, that I would as a courtesy to other developers, because the deprecate calls are not exactly self explanatory. Long story short, I think that for the ones that breach multiple lines anyway, I would use the keywords to make it easier to read.

This also helps a little with making sure that the message, which may be multi-line, is easier to separate from the next arguments. Consider how small the difference is between these two:
```python
@deprecate(
    "0.1",
    "At present, ixdat measurements have a `calibration_list` but no compound "
    "`calibration`, and the property just returns the first from the list.",
    None,
)

@deprecate(
    "0.1",
    "At present, ixdat measurements have a `calibration_list` but no compound ",
    "`calibration`, and the property just returns the first from the list.",
    None,
)
```
where the latter is a bug that fills in a nonsensical `hard_deprecation_release`. Whereas if we do:
```python
@deprecate(
    last_supported_version="0.1",
    update_message=(
        "At present, ixdat measurements have a `calibration_list` but no compound "
        "`calibration`, and the property just returns the first from the list."
    )
    hard_deprecation_release=None,
)
```
that risk is seriously mitigated.

I will not comment on this more than once.

KennethNielsen commented on src/ixdat/measurements.py:
```diff
@@ -589,9 +610,25 @@ def __getitem__(self, key):
+    def _cache_series(self, key, series):
+        """Cache `series` such that it can be looked up with its name or with `key`."""
+        self._cached_series[key] = series  # now it can be looked up with by `key`
+        # If the name of the series is not `key`, we can get in a situation where
```
I'm a little in doubt about this change. Is this change changing what aliases may or may not be used for? I was under the assumption that aliases were either defined by readers or calibrations and were hence forth static, but this certainly changes that. Is that on purpose? And/or am I understanding the concept of alias wrong? Additionally, how would this compare to simply adding the series to `_cached_series` under their name also? Or will that somehow mess up because because of a possibility for more than one with the same name?

ScottSoren replied:
Good point. I hadn't thought of `aliases` as static, but I hadn't thought of it as  necessarily dynamic either. No reason to make it dynamic when there's another option. simple better than complex better than complicated, and `cached_series` is clearly already dynamic.
I've changed it to your suggestion.

KennethNielsen commented on src/ixdat/plotters/ec_plotter.py:
```diff
@@ -12,18 +13,24 @@ def __init__(self, measurement=None):
         """Initiate the ECPlotter with its default Meausurement to plot"""
         self.measurement = measurement
 
+    @deprecate("0.1", "Use `v_name` instead.", "0.3", kwarg_name="V_str")
```
Ha. I had never, annoyingly, consider how these would stack. Presumably the `@wraps` decorator copies the necessary callable signature over, otherwise it wouldn't work. Have you tested that it works both for more than the first argument name "j_color", because I don't test that.

KennethNielsen commented on src/ixdat/plotters/ec_plotter.py:
```diff
@@ -12,18 +13,24 @@ def __init__(self, measurement=None):
         self.measurement = measurement
 
+    @deprecate("0.1", "Use `v_name` instead.", "0.3", kwarg_name="V_str")
+    @deprecate("0.1", "Use `j_name` instead.", "0.3", kwarg_name="J_str")
```
Here BTW we might excuse ourselves from writing the kwarg names out to keep it on one line.

KennethNielsen commented on src/ixdat/plotters/ec_plotter.py:
```diff
@@ -37,18 +44,20 @@ def plot_measurement(
-            J_str (str): DEPRECIATED. now j_name
             v_color (str): The color to plot v_name. Defaults to black.
             j_color (str): The color to plot j_name. Defaults to red.
+            V_str (str): DEPRECATED
```
Maybe, you should keep the new names here still, so that it is deducible from the API docs only, without running the code.

KennethNielsen commented on src/ixdat/plotters/ecms_plotter.py:
```diff
@@ -12,6 +15,13 @@ def __init__(self, measurement=None):
         self.ec_plotter = ECPlotter(measurement=measurement)
         self.ms_plotter = MSPlotter(measurement=measurement)
 
+    @deprecate("0.1", "Use `v_name` instead.", "0.3", kwarg_name="V_str")
```
:eyes: "Let's deprecate" :notes:

KennethNielsen commented on src/ixdat/plotters/ecms_plotter.py:
```diff
@@ -64,13 +81,18 @@ def plot_measurement(
                 Defaults to the name of the series `measurement.current`
             v_color (str): The color to plot the variable given by 'V_str'
             j_color (str): The color to plot the variable given by 'J_str'
+            V_str (str): DEPRECATED
```
Maybe also here, list the new names

KennethNielsen commented on src/ixdat/techniques/cv.py:
```diff
@@ -324,6 +325,15 @@ def diff_with(self, other, v_list=None, cls=None, v_scan_res=0.001, res_points=1
 
+class CyclicVoltammagram(CyclicVoltammogram):
+
+    # FIXME: decorating the class itself doesn't work because the callable returned
```
Wow. Had not considered that. Nice catch.

KennethNielsen replied:
The reason I was angling to avoid this, was that I was unsure if the actual class name was being used anywhere in (de)/serialization code and so would cause a data hygiene problem. If that is the case, we might just move ahead and hard deprecate immediately.

KennethNielsen replied:
I will raise an issue on it, since I've documented being able to use it for the class.

KennethNielsen commented on src/ixdat/techniques/ms.py:
```diff
@@ -145,7 +150,11 @@ def grab_flux(
                 If not given, no background is subtracted.
             remove_background (bool): Whether to remove a pre-set background if available
                 Defaults to True.
+            removebackground (bool): DEPRECATED
```
and here, new name. I will not comment on it further.

KennethNielsen reviewed (changes requested):
Looks great. Nice to see that decorator being put to good use.

There are a few typos and some questions in the review. Mostly simple stuff I think. Should be a quick turn around. Just let me know whether any of it warrants a second look.

AnnaWiniwarter reviewed (changes requested):
Of the edited files, I only had a look at NEXT_CHANGES.rst (see comments there, Kenneth had more than I did).
There is no mention of the change to calling calibrated molecules now with `mol_list=["CO2@M44"]` instead of old `mol_list=["CO2_M44"]` (and this just results in an error instead of deprecation comment)
Also,   `co_strip.add_calibration(RE_vs_RHE=0, A_el=0.196, R_Ohm=0)` doesn't work, but I cannot really figure out what the new syntax should be if I dont use `.calibration()` anymore. so I couldn't test if issue #29 is solved

Then I tried running some of my old scripts where I encountered errors in the calibration.
First of all, somehow the importing of .csv files exported with some 0.1.x version seems to be broken: 
![image](https://user-images.githubusercontent.com/96479350/159511116-7447fe92-77fe-4841-a2a8-9719a5490fa0.png)
I did not investigate this further, but I think I already ran into this problem when using the new "main" branch after The Big Merge.

If I then import the data just straight from the Zilien file instead and try and do an ECMS calibration, it still breaks when I try to use `.as_cv()` to convert it to an ECMS_Voltammogram object. Instead of choosing one cycle if I call `my_cv[1]` it chooses some part of the entire dataset, and and a different part `my_cv[2]` but both kind of stretching over the entire time of the measurement. 
on the plus side, the result is the same whether I call `.as_cv()` before or after the calibration. That means that issue #22 appears to be solved (it is now possible to use ecms_calibration on a ECMSCyclicVoltammogram object)

finally, I tried to convert a combination of files imported from Zilien and biologic data to a CyclicVoltammogram object (issue #20). this works, but when plotting it somehow plots the current double with a weird time shift: 
![Capture2](https://user-images.githubusercontent.com/96479350/159525267-785c1362-d6e7-4471-8f7a-729ba85b669c.PNG)
(top image is if I import full dataset from zilien file, bottom if I combine the MS data from zilien (imported with `MSMeasurement.read()` with EC data from biologic. Note this happens regardless of whether I convert it to a ECMSCyclicVoltammogram object or not.

Hope my comments are somewhat clear.

ScottSoren commented on NEXT_CHANGES.rst:
```diff
@@ -0,0 +1,102 @@
+changes and additions to ixdat's API associated with your contribution. The idea is
+that what you write here informs other developers what is on its way and then will be
+copied to CHANGES.rst when the next version of ixdat is distributed. Please include
+links to relevant Issues, Discussions, and PR's on github.
```
Done.

ScottSoren commented on NEXT_CHANGES.rst:
```diff
@@ -0,0 +1,102 @@
+
+  - ``v_name`` (replaces ``V_str``)
+  - ``j_name`` (replaces ``J_str``)
+  - ``E_name`` (replaces ``E_str``)
```
For code that should represent something a scientist could conceivably write on paper or on the blackboard, I want `ixdat` to always stay as close as possible to what they'd write. I'd never write "e" for potential or "i" for current (I have seen the latter though). I guess "E" and "I" carry over from the biologic names for the raw data. Calling normalized current "j"  (lowercase) is pretty standard, I think. "v" for calibrated potential might be less standardized (E, V, and U are maybe more common) but capital "V" clashes with the symbol for the unit Volt. Hm... now I kind of regret not calling it "U". what do you think? It could be "U" and "J" for calibrated potential and current, to keep them all uppercase. What do you think? Now's probably the last chance to change them, so it should be something we agree on.

AnnaWiniwarter replied:
I like "U" much more than "v". Didn't you use "U" also in the figure captions in EC_MS?

AnnaWiniwarter replied:
oh, and yes, I guess upper case "J" would make more sense for consistency. though the lower case letter is what one would use for current density on the blackbord, isn't it?

ScottSoren replied:
Yeah, typically ``v_name`` is "U vs RHE / [V]"   (same in EC_MS though it was called ``V_str``)
So... yeah lets do "J_name" and "U_name" then. Tribalism for the win. Real electrochemists call potential "U", haha ;) 

On it. Bleh, more refactoring.

ScottSoren replied:
was just reminded when opening ec.py to start that ``j_name`` is typically "J / [mA cm$^{-2}$]", haha. 
"E", "I", "U", and "J" are then least confusing.

ScottSoren replied:
Oh, but there's more:
``ec_meas.v`` and ``ec_meas.j`` returns the potential and current. Should that be ``ec_meas.U`` and ``ec_meas.J`` ?
And aren't cyclic voltammograms and linear sweeps usually plotted as "J-V curves"?

AnnaWiniwarter replied:
Yes, I guess it should be `ec_meas.U` and `ec.meas.J`, even though when writing with this syntax it feels more like code and less like electrochemistry, so the capitalization seems a bit off. and yes, they are called J-V curves, but I with `U_name` being "U vs RHE / [V]" I think having the U there is the least confusing, as it draws attention to the naming convention in the plot. if someone doesn't like the U there, they are warned and can change it to V manually.

ScottSoren commented on NEXT_CHANGES.rst:
```diff
@@ -0,0 +1,102 @@
+
+- Generalization of multiple calibrations:
+
+  ``Measurement.calibration`` is deprecated in favor of ``Measurement.calibration_list``.
```
It's a getter and a setter... but I think one thing that deprecated. Unless you have a better way to phrase this, I'll leave it be. I expect there will be a development on calibration handling during ixdat v0.2.x, so we can come back to it.

KennethNielsen replied:
I would suggest something along the lines of "The single calibration gettter and better ... has been deprecated in favor of the ... getter, which will return the current list of calibrations and ... which is used to add a calibration to that list." possibly expanded with "This change is to accommodate the idea of multiple calibrations for a measurement which is used to ..."

Written on my phone, so I will forego typing out the property names 😁

ScottSoren replied:
Done. and more.

ScottSoren commented on NEXT_CHANGES.rst:
```diff
@@ -0,0 +1,102 @@
+  MS data, ``axes[1]`` electrode potential, ``axes[2]=None`` and ``axes[3]`` being the
+  electrode current.
+
+- Renamed key-word arguments in EC, EC-MS, and SEC plotting functions:
```
replaced.

ScottSoren commented on NEXT_CHANGES.rst:
```diff
@@ -0,0 +1,102 @@
+- Carrying calibrations over through measurement combination (hyphenation or appending
+  with the `+` operator) or transformation (through copying and technique-changing
+  functions like `ECMeasurement.as_cv()`. This is done via replacing the use of a single
+  ``Measurement.calibration`` with a ``calibration_list`` which can be appended,
```
You did review the code that does it ;)

ScottSoren commented on docs/source/developing.rst:
```diff
@@ -15,15 +15,15 @@ Git and Github
-
 To develop ixdat, you will need to use git and github. This means
 
 - `install git <https://git-scm.com/downloads>`_. Git bash is strongly recommended for Windows
```
the word strongly is removed.

ScottSoren commented on docs/source/developing.rst:
```diff
@@ -40,14 +40,18 @@ To develop ixdat, you will need to use git and github. This means
-    git branch my_branch_name
-    git checkout my_branch_name
+    git branch my_feature_branch_name
+    git switch my_feature_branch_name
```
The line above it creates the branch.
But that's nice. I'll combine the two commands.

ScottSoren commented:
Hi @KennethNielsen and @AnnaWiniwarter , 

I've implemented the changes requested on this PR. This includes describing "mol@mass" and giving an example of ``add_calibration()`` useage in NEXT_CHANGES. This also includes, renaming "v_name" to "U_name" and "j_name" to "J_name" to match what Anna and I feel most natural on the blackboard, keep all the raw and calibrated current and potential shortcuts uppercase, and not clash with the useage of "v" to mean general "value" in ixdat. 

Huge thanks to Anna for testing. I had tested against the usual suspects (tutorials and article repositories) but hadn't found all the bugs. As briefly discussed with Anna on slack, this probably means two more PR's tomorrow before ixdat v0.2.0 release:

- back-compatability with .csv files exported by ixdat v0.1.x
- debugging something about calibration carryover

Let me know if this one is in your eyes good to merge now.

KennethNielsen commented on NEXT_CHANGES.rst:
```diff
@@ -0,0 +1,102 @@
+
+  The old property names are deprecated.
+
+- Renamed keyward in ``MSMeasurement.plot_flux()``:
```
it appears to still be spelled wrong

KennethNielsen reviewed (approved):
I've left a few comments on discussions and unresolved a single discussion  around a typo. They are all minor and so once you have had a look at the comments I consider this good to go.

AnnaWiniwarter commented:
I think it would be great to also have the new syntax for an EC calibration in NEXT_CHANGES, unless you think it shouldn't go in there because `ec_data.calibrate(E_vs_RHE=1, A_el=0.5) ` still works. Else I think it's fine to merge.

ScottSoren commented:
Alright, @AnnaWiniwarter , I think I have addressed all your concerns here. What was done since last review:

- Renamed `v_name` to `U_name`, `j_name` to `J_name`, etc, updated also where I could find it in documentation, deprecation decorators, and variable names within functions. 
- Generalized the `calibrate` method up from `ECMeasurement` to `Measurement` so that you can use `calibrate` for any type of measurement as a shortcut (rather than `add_calibration`). 
- Described and gave examples of `calibrate` in `NEXT_CHANGES.rst`
- Described `mol@mass` notation with an example in `NEXT_CHANGES.rst` (should go in ms documentation too)
- Described the workflow of using  `CyclicVoltammogram` with an example in [docs/techniques/electrochemistry](https://github.com/ixdat/ixdat/blob/lets_deprecate/docs/source/technique_docs/electrochemistry.rst)
- Ensured that future .csv exports state the ixdat version in the top line
- Added documentation of how to read old .csv's in NEXT_CHANGES.rst

Let me know if all is good! If so I think this is indeed the last PR to v0.2.0 as hoped :) 

@KennethNielsen , since there were some non-trivial changes resulting from Anna's testing, you may also want to give another quick glance. 

Looking forward to merging this!

AnnaWiniwarter commented:
Looks great! The only comment I have is that in the documentation you now recommend to use redefine cycle, but you don't explain why cycles are messed up if you don't use "redefine cycle" like you wrote me earlier:  "Before you've used this, the` counter cycle` is the "cycle number" column of the imported data. If you use a loop, this repeats itself, so you can get chucks of disconnected data when selecting for it. If you run `redefine_cycle` , the` counter cycle` is redefined so each cycle has a unique number." 
But with the recommendation and the example you give it should be fine to just leave it like it is, since now there is a guide stating that it's a good idea to use it.

KennethNielsen commented on NEXT_CHANGES.rst:
```diff
@@ -0,0 +1,200 @@
+    >>> my_ecms_meas.calibrate(
+    >>>     RE_vs_RHE=0.715, A_el=0.196, ms_cal_results=[my_H2_at_M2, my_O2_at_M32]
+    >>> )   # note that all of these are keyword arguments to ECMSCalibration.
+    >>> my_ecms_meas.plot(mol_list=["H2", "O2"])
```
Looks great and I understand the mechanism better now. In respect to my question the answer then is, if I understand it correct, that concatenating measurements with different calibrations makes no sense, so we will simple use the one that ends up the latest for any given parameter

KennethNielsen commented on NEXT_CHANGES.rst:
```diff
@@ -0,0 +1,200 @@
+
+ixdat.readers
+^^^^^^^^^^^^^
+- ixdat-eported .csv compatability
```
some typo here in eported

ScottSoren replied:
exported.

KennethNielsen commented on NEXT_CHANGES.rst:
```diff
@@ -0,0 +1,200 @@
+  For even earlier .csv files (exported by ixdat version <0.1.5), you will also need to
+  specify the technique.
+
+  Starting in ixdat v0.2.0, ixdat-exported files will have the ixdat version in the header.
```
Good idea

KennethNielsen commented on docs/source/developing.rst:
```diff
@@ -40,14 +41,17 @@ To develop ixdat, you will need to use git and github. This means
+    git switch branch_to_work_from
 
-    git checkout branch_to_work_from
+  Usually it makes most sense to work from the main branch::
```
Maybe the formulation should be "Usually ... to **start** work from ..."

KennethNielsen replied:
so as to not cause confusion below when you fire up the feature branch

ScottSoren replied:
Usually it makes most sense to start work by branching from the main branch

KennethNielsen commented on docs/source/technique_docs/electrochemistry.rst:
```diff
@@ -4,22 +4,22 @@ Electrochemistry
 
 The main TechniqueMeasurement class for electrochemistry is the ``ECMeasurement``.
-Subclasses of ``ECMeasurement`` include ``CyclicVoltammagram`` and ``CyclicVoltammagramDiff``.
+Subclasses of ``ECMeasurement`` include ``CyclicVoltammogram`` and ``CyclicVoltammogramDiff``.
```
Don't remember if I saw a deprecation on the mis-spelled Diff as well

KennethNielsen commented on docs/source/technique_docs/electrochemistry.rst:
```diff
@@ -29,6 +29,46 @@ the following Tutorials:
 
+
+``CyclicVoltammogram`` inherits from ``ECMeasurement`` and adds useful functionality for
+cyclic voltammotry analysis. This includes:
```
"..mmEtry andlysis..."

ScottSoren replied:
that word can go spell itself.

KennethNielsen commented on docs/source/technique_docs/electrochemistry.rst:
```diff
@@ -29,6 +29,46 @@ the following Tutorials:
+  >>> ax = my_cv[2].plot(color="b", label="earlier")
+  >>> my_cv[50].plot(ax=ax, color="r", label="later")
+  >>> ax.legend()
+  >>> ax.get_figure().savefig("two_cycles.png")
```
that is beautiful API wise

KennethNielsen commented on src/ixdat/exporters/csv_exporter.py:
```diff
@@ -84,6 +85,8 @@ def prepare_header_and_data(self, measurement, v_list, tspan):
             s_list.append(v_name)
 
         header_lines = []
+        ixdat_version_line = f"ixdat version = {__version__}\n"
```
You can also choose to version the file format, depending on what is most convenient

ScottSoren replied:
this is most convenient.

KennethNielsen commented on src/ixdat/plotters/ec_plotter.py:
```diff
@@ -37,18 +42,19 @@ def plot_measurement(
+            J_color (str): The color to plot J_name. Defaults to red.
+            V_str (str): DEPRECATED. Use `U_name`.
+            J_str (str): DEPRECATED. Use `J_name`.
+            V_color (str): DEPRECATED. Use `U_color`.
```
Is that "Use `v_color`"? At least that's what's in the decorator

ScottSoren replied:
`v_color` was a rejected intermediate between `V_color` and `U_color` which never saw PyPI. (Same with `V_name`.) Decorator fixed.

KennethNielsen commented on src/ixdat/plotters/ec_plotter.py:
```diff
@@ -58,39 +64,45 @@ def plot_measurement(
+        U_name = U_name or V_str
+        J_name = J_name or J_str
+        U_color = U_color or V_color
+        J_color = J_color or J_color
```
Is there missing an combination with old arg here for `v_color`?

KennethNielsen reviewed (approved):
I noticed a few more typos, which should of course be fixed, but I will mark as approved, so you won't have to wait for me before merge.

ScottSoren commented:
Thanks @AnnaWiniwarter and @KennethNielsen for your reviews and tests of this! 
Merging now to main. I'll then change the version to 0.2.0 (removing the "dev"), copy the contest of NEXT_CHANGES.rst to CHANGES.rst, and then upload the main branche's docs to ixdat.readthedocs.org and its packaged code to pypi.org/project/ixdat
